File: ScriptGraph/BuiltInLua/CGSetPropertyValue.lua
local CGSetPropertyValue = CGSetPropertyValue or {}
CGSetPropertyValue.__index = CGSetPropertyValue

function CGSetPropertyValue.new()
    local self = setmetatable({}, CGSetPropertyValue)
    self.inputs = {}
    self.outputs = {}
    self.nexts = {}
    self.object = nil
    self.property = nil
    self.propertyFunc = nil
    self.resource = nil
    self.resourcePath = nil
    return self
end

function CGSetPropertyValue:setNext(index, node)
    self.nexts[index] = node
end

function CGSetPropertyValue:setInput(index, func)
    self.inputs[index] = func
end

function CGSetPropertyValue:getOutput(index)
    return self.outputs[index]
end

function CGSetPropertyValue:execute()
    if self.object == nil or self.property == nil or self.propertyFunc == nil then
        return
    end
    Amaz.LOGE("AE_LUA_TAG", "property:"..self.valueType)
    local inputValue = self.inputs[1]();
    if inputValue == nil then
        if self.valueType == 'Mesh' or self.valueType == 'Material' or self.valueType == 'Texture2D' then
            inputValue = self.resource
        end
    end
    local objects = {}
    objects[1] = self.object
    self.propertyFunc:setProperty(objects, self.property, inputValue, self.valueType)
    if self.nexts[0] then
        self.nexts[0]()
    end
    return true
end

return CGSetPropertyValue

File: ScriptGraph/BuiltInJS/CGSetPropertyValue.js
/**
 * @file CGSetPropertyValue.js
 * @author
 * @date 2021/8/24
 * @brief CGSetPropertyValue.js
 * @copyright Copyright (c) 2021, ByteDance Inc, All Rights Reserved
 */

const Amaz = effect.Amaz;
const {BaseNode} = require('./BaseNode');

class CGSetPropertyValue extends BaseNode {
  constructor() {
    super();
    this.objects = [];
    this.property = null;
    this.resource = null;
    this.resourcePath = null;
    this.propertyFunc = null;
    this.valueType = null;
  }

  execute(index) {
    if (
      this.objects.length === 0 ||
      this.property === null ||
      this.inputs[1] === undefined ||
      this.propertyFunc === null
    ) {
      return false;
    }

    let inputValue = this.inputs[1]();
    if (inputValue === null) {
      if (this.valueType === 'Mesh' || this.valueType === 'Material' || this.valueType === 'Texture2D') {
        inputValue = this.resource;
      }
    }

    this.propertyFunc.setProperty(this.objects, this.property, inputValue, this.valueType);
    if (this.nexts[0]) {
      this.nexts[0]();
    }
    return true;
  }
}

exports.CGSetPropertyValue = CGSetPropertyValue;
